test/test_nn.py

```diff
@@ -16456,6 +16456,118 @@ def _helper(zero_infinity):
+        targets = torch.randint(low=1, high=vocab_size - 1, size=(batch_size, target_length),
+                                dtype=torch.int, device=device)
+        input_lengths = batch_size * [input_length]
+        target_lengths = batch_size * [target_length]
+
+        log_probs_no_bd = log_probs.squeeze(1).detach().clone().requires_grad_()
+        targets_no_bd = targets.squeeze(0).detach().clone()
+        input_lengths_no_bd = torch.tensor(input_length)
+        target_lengths_no_bd = torch.tensor(target_length)
+
+        log_probs_refs = [log_probs.detach().clone().requires_grad_() for _ in range(4)]
+        log_probs_no_bd_refs = [log_probs_no_bd.detach().clone().requires_grad_() for _ in range(2)]
+
+        func_ctc = torch.nn.functional.ctc_loss
+        ctc_loss = nn.CTCLoss(reduction=reduction, zero_infinity=True)
+
+        losses = []
+        losses_no_bd = []
+
+        # testing the functional version
+        # batched case. log_probs.shape = (T, N, C), targets = (N, S), input_lengths/target_lengths = (N,)
+        losses.append(func_ctc(log_probs_refs[0], _targets_device(targets),
+                               input_lengths, target_lengths, reduction=reduction, zero_infinity=True))
+        # batched case. input.shape = (T, N, C), targets = (S,), input_lengths/target_lengths = (N,)
+        losses.append(func_ctc(log_probs_refs[1], _targets_device(targets_no_bd),
+                               input_lengths, target_lengths, reduction=reduction, zero_infinity=True))
+        # unbatched case. input.shape = (T, C), targets = (S,), input_lengths/target_lengths = (N,)
+        losses_no_bd.append(func_ctc(log_probs_no_bd_refs[0], _targets_device(targets_no_bd),
+                                     input_lengths_no_bd, target_lengths_no_bd, reduction=reduction,
+                                     zero_infinity=True))
+
+        # same tests as above but with the module instead of functional version
+        losses.append(ctc_loss(log_probs_refs[2], _targets_device(targets), input_lengths, target_lengths))
+        losses.append(ctc_loss(log_probs_refs[3], _targets_device(targets_no_bd), input_lengths, target_lengths))
+        losses_no_bd.append(ctc_loss(log_probs_no_bd_refs[1], _targets_device(targets_no_bd),
+                                     input_lengths_no_bd, target_lengths_no_bd))
+
+        for loss in losses + losses_no_bd:
+            loss.backward()
+
+        return losses, losses_no_bd, log_probs_refs, log_probs_no_bd_refs
+
+    def _assertEqual_list(self, expected, list_to_compare, atol=None, rtol=None):
+        for ele in list_to_compare:
+            self.assertEqual(expected, ele, atol=atol, rtol=rtol)
+
+    @parametrize_test("reduction", ['none', 'mean', 'sum'])
+    def test_CTCLoss_no_batch_dim(self, device, reduction):
+        input_length = 40
+        vocab_size = 3
+        target_length = 12
+
+        # cuda/cudnn case
+        if torch.cuda.is_available():
```

Comment:
To simplify the device placement logic, I think you can just do this:
1. Use `device=device` for both input and target
2. If torch.cuda.is_available() and torch.backends.cudnn.enabled, move target to CPU (i.e. `target = target.cpu()`)
3. Do function / module calls with input + target
4. Do checks

As another possibly simplifying thing, you could also do:
```
@parametrize_test("use_module_form", [True, False])
```

and call the functional form if `use_module_form=False`